presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -1366,27 +1372,32 @@ private boolean canPushdownPartialMergeThroughLowMemoryOperators(PlanNode node)
                 .allMatch(this::canPushdownPartialMergeThroughLowMemoryOperators);
     }
 
-    private static Map<Symbol, Symbol> computeIdentityTranslations(Assignments assignments)
+    public static Map<VariableReferenceExpression, VariableReferenceExpression> computeIdentityTranslations(Assignments assignments, TypeProvider types)
     {
-        Map<Symbol, Symbol> outputToInput = new HashMap<>();
-        for (Map.Entry<Symbol, Expression> assignment : assignments.getMap().entrySet()) {
+        Map<VariableReferenceExpression, VariableReferenceExpression> outputToInput = new HashMap<>();
+        for (Map.Entry<VariableReferenceExpression, Expression> assignment : assignments.getMap().entrySet()) {
             if (assignment.getValue() instanceof SymbolReference) {
-                outputToInput.put(assignment.getKey(), Symbol.from(assignment.getValue()));
+                outputToInput.put(assignment.getKey(), toVariableReference(Symbol.from(assignment.getValue()), types));
             }
         }
         return outputToInput;
     }
 
+    public static VariableReferenceExpression toVariableReference(Symbol symbol, TypeProvider types)
```

Comment:
It might be better placed in a place that is for common RowExpression utilities, like `Expressions`.

Reply:
It was supposed to be temporary so I wasn't too picky about it (though I think this one stays till the end of this PR, but by the end it was only used in partition related classes). I can add another commit to move this but moving this is an early commit will cause too much trouble in rebasing. What do you think?

Reply:
Add an additional commit sounds great :). Having this utility will make my life easier :) but I agree this eventually will be gone after everything is converted.